File: golang/2641.cousins-in-binary-tree-ii.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func replaceValueInTree(root *TreeNode) *TreeNode {
	if root == nil {
		return root
	}

	q := []*TreeNode{root}
	prevSum := root.Val

	for level := 0; len(q) > 0; level++ {
		levelSum := 0

		for size := len(q); size > 0; size-- {
			node := q[0]
			q = q[1:]

			node.Val = prevSum - node.Val

			brothers := 0
			if node.Left != nil {
				brothers += node.Left.Val
			}
			if node.Right != nil {
				brothers += node.Right.Val
			}

			if node.Left != nil {
				levelSum += node.Left.Val
				node.Left.Val = brothers
				q = append(q, node.Left)
			}

			if node.Right != nil {
				levelSum += node.Right.Val
				node.Right.Val = brothers
				q = append(q, node.Right)
			}
		}

		prevSum = levelSum
	}

	return root
}

File: python/2641.cousins-in-binary-tree-ii.py

import operator
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def replaceValueInTree(self, root: TreeNode | None) -> TreeNode | None:
        if not root:
            return root

        q = deque([(root, root.val)])

        level = 0
        n = operator.itemgetter(0)
        v = operator.attrgetter("val")

        while q:

            sumLevel = sum(map(v, map(n, q)))

            for _ in range(len(q)):
                node, brothers = q.popleft()

                node.val = sumLevel - brothers

                brothers = 0
                if node.left:
                    brothers += node.left.val
                if node.right:
                    brothers += node.right.val

                if node.left:
                    q.append((node.left, brothers))

                if node.right:
                    q.append((node.right, brothers))

            level += 1

        return root


